keps/sig-node/4639-oci-volume-source/README.md

```diff
@@ -0,0 +1,940 @@
+
+The volume-populators API extension allows you to populate a volume with data from an external data source when the volume is created.
+This is a good solution for restoring a volume from a snapshot or initializing a volume with data from a database backup. However, it does not
+address the desire to use OCI distribution, versioning, and signing for mounted data.
+
+The proposed in-tree OCI VolumeSource provides a direct and integrated approach to mount OCI artifacts, leveraging the existing OCI
+infrastructure for packaging, distribution, and security.
+
+### Custom CSI Plugin
+
+See [https://github.com/warm-metal/container-image-csi-driver](https://github.com/warm-metal/container-image-csi-driver)
+
+An out-of-tree CSI plugin can provide flexibility and modularity, but there are trade-offs to consider:
+
+ - Complexity of managing an external CSI plugin. This includes handling the installation, configuration, and updates of the CSI driver, which adds
+   an additional operational burden. For a generic, vendor-agnostic, and widely-adopted solution this would not make sense.
+ - Supporting the image pull secrets as well as credentials provider will be tricky and needs to be reimplemented with the separate API calls.
+ - External CSI plugins implement their own lifecycle management and garbage collection mechanisms,
+   yet these already exist in-tree for OCI images.
+ - Performance: There is additional overhead with an out-of-tree CSI plugin, especially in scenarios requiring frequent image pulls
+   or large volumes of data.
+
+### Advantages of In-Tree OCI VolumeSource
+
+1. **Leverage Existing Mechanisms:**
+   - **No New Data Types or Objects:** OCI images are already a core part of the Kubernetes ecosystem. Extending support for OCI artifacts, many of
+     the same mechanisms will be reused. This ensures consistency and reduces complexity, as both adhere to the same OCI image format.
+   - **Existing Lifecycle Management and Garbage Collection:** Kubernetes has efficient lifecycle management and garbage collection mechanisms for
+	 volumes and container images. The in-tree OCI VolumeSource will utilize these existing mechanisms.
+
+2. **Integration with Kubernetes:**
+   - **Optimal Performance:** Deep integration with the scheduler and kubelet ensures optimal performance and
+     resource management. This integration allows the OCI VolumeSource to benefit from all existing optimizations and features.
+   - **Unified Interface:** Users interact with a consistent and unified interface for managing volumes, reducing the learning curve and
+     potential for configuration errors.
+
+3. **Simplified Maintenance and Updates:**
+   - **Core Project Maintenance:** In-tree features are maintained and updated as part of the core project. It makes sense
+     for widely-used and vendor agnostic features to utilize the core testing infrastructure, release cycles, and security updates.
```

Comment:
We have widely-used and vendor agnostic CSI drivers under Kubernetes umbrella - NFS, SMB, and shared-secrets. They use the core testing infrastructure. They can have different cycles and security updates. if they're not good enough, we should fix those processes for all non-in-tree components, not merge everything back to k/k.

Reply:
“Core Project Maintenance” can sometimes read like: if we get this merged, someone will _have_ to maintain it.